### ISSUE TITLE
Mathematical classification of solvable layouts

### ISSUE THREAD

**Issue by tkoeppe** [tkoeppe/lightgame#1]
@smarnach: Can you provide conditions that determine whether a layout is solvable, or even enumerate solutions?

**Comment by smarnach** [comment]
I don't think there is a simple criterion for solvability. The definition of solvability is inherently recursive, and I can't see any invariants that would allow to simplify the definition significantly.

The algorithm could certainly be optimised, though. One possible optimisation is to find a way to efficiently detect in each step whether switching on a light divides the remaining space into two connected components. For this detection to be efficient, it would need to happen locally, and this would require some preprocessing of the topology. Essentially, you need to detect the "genus" of the topology in before starting, and reduce it by one whenever you touch one of the "holes". The details are a bit involved, though, and don't have time to work it out.

**Comment by tkoeppe** [comment]
Thanks a lot for your input!

I've thought about connectedness, too, but I figure that the ratio of layouts that are obstructed topologically is small, and so that particular analysis (which seems non-local by definition) does not seem worthwhile. Thanks, though!

**Comment by smarnach** [comment]
It's not only about testing connectedness upfront. The optimisation I propose is to check whether the remaining space has _become_ disconnected during the brute-force solution. This would allow to exit branches early that don't have a chance of succeeding, and I'd reckon it will significantly speed up the test for bigger boards.

**Comment by tkoeppe** [comment]
But if you spray new blocked tiles randomly at a large board, it's very very unlikely that you'll disconnect a part, isn't it? That is, as you're randomly searching for solvable boards, _most_ candidates will not be disconnected?

There is of course some value in this analysis: at the moment, you can start with a disconnected (manually created) layout and hit "augment" (with an augment count smaller than the number of empty tiles in each component), and the program will be in an infinite loop, since it keeps trying new layouts until it finds one that works.

**Comment by smarnach** [comment]
Sorry for being unclear. I'm only talking about optimising `Game::IsSolvable()`, specifically the `solve_one` closure. If it becomes faster to decide whether a board is solvable, it also becomes faster to generate a solvable one. In `solve_one` it is very common to divide the remaining space, but the current algorithm will keep backtracking into both parts, which takes a lot of time for bigger boards.

**Comment by tkoeppe** [comment]
Oh, I see what you mean: during the exploration, the current partial path may well end up disconnecting the remaining space, which would be an instant dead end, right?

**Comment by smarnach** [comment]
Yes, exactly, that's what I mean. In isn't _trivial_ to check this locally in each step, but I think it's possible with some light-weight preprocessing, as explained in my first comment.

**Comment by tkoeppe** [comment]
@smarnach Could you contact me by email perhaps?
